### PR TITLE
core: Try to prevent infinite recursive template instantiation

### PR DESCRIPTION
**Pull request by daandemeyer**
To prevent situations like in #17602 from happening, let's drop
direct recursive template dependencies. These will almost certainly
lead to infinite recursion so let's drop them immediately to avoid
instantiating potentially thousands of irrelevant units.

Example of a template that would lead to infinite recursion which
is catched by this check:

notify@.service:

```
[Unit]
Wants=notify@%n.service
```

Results when booting a mkosi image that includes the reproduce from @17602 (global onfailure handler)

```
Spawning container image on /home/daandemeyer/projects/systemd/image.raw.
Press ^] three times within 1s to kill container.
systemd mkosi running in system mode (+PAM +AUDIT -SELINUX -APPARMOR +IMA +SMACK +SECCOMP +GCRYPT +GNUTLS +OPENSSL +ACL +BLKID +CURL +ELFUTILS -FIDO2 +IDN2 -IDN +IPTC +KMOD +LIBCRYPTSETUP +LIBFDISK +PCRE2 -PWQUALITY +P11KIT +QRENCODE +BZIP2 +LZ4 +XZ +ZLIB +ZSTD +XKBCOMMON +UTMP +SYSVINIT default-hierarchy=unified)
Detected virtualization systemd-nspawn.
Detected architecture x86-64.

Welcome to Arch Linux!

Initializing machine ID from container UUID.
issue-17602@systemd-journald.service.service: Dropping dependency issue-17602@issue-17602@systemd-journald.service.service.service that would lead to infinite recursion
issue-17602@systemd-tmpfiles-clean.service.service: Dropping dependency issue-17602@issue-17602@systemd-tmpfiles-clean.service.service.service that would lead to infinite recursion
issue-17602@systemd-remount-fs.service.service: Dropping dependency issue-17602@issue-17602@systemd-remount-fs.service.service.service that would lead to infinite recursion
issue-17602@systemd-fsck-root.service.service: Dropping dependency issue-17602@issue-17602@systemd-fsck-root.service.service.service that would lead to infinite recursion
issue-17602@emergency.service.service: Dropping dependency issue-17602@issue-17602@emergency.service.service.service that would lead to infinite recursion
issue-17602@rescue.service.service: Dropping dependency issue-17602@issue-17602@rescue.service.service.service that would lead to infinite recursion
issue-17602@initrd-switch-root.service.service: Dropping dependency issue-17602@issue-17602@initrd-switch-root.service.service.service that would lead to infinite recursion
issue-17602@shadow.service.service: Dropping dependency issue-17602@issue-17602@shadow.service.service.service that would lead to infinite recursion
issue-17602@systemd-sysusers.service.service: Dropping dependency issue-17602@issue-17602@systemd-sysusers.service.service.service that would lead to infinite recursion
issue-17602@systemd-update-done.service.service: Dropping dependency issue-17602@issue-17602@systemd-update-done.service.service.service that would lead to infinite recursion
issue-17602@systemd-udevd.service.service: Dropping dependency issue-17602@issue-17602@systemd-udevd.service.service.service that would lead to infinite recursion
issue-17602@systemd-hwdb-update.service.service: Dropping dependency issue-17602@issue-17602@systemd-hwdb-update.service.service.service that would lead to infinite recursion
issue-17602@systemd-networkd.service.service: Dropping dependency issue-17602@issue-17602@systemd-networkd.service.service.service that would lead to infinite recursion
issue-17602@systemd-user-sessions.service.service: Dropping dependency issue-17602@issue-17602@systemd-user-sessions.service.service.service that would lead to infinite recursion
issue-17602@systemd-update-utmp-runlevel.service.service: Dropping dependency issue-17602@issue-17602@systemd-update-utmp-runlevel.service.service.service that would lead to infinite recursion
issue-17602@systemd-update-utmp.service.service: Dropping dependency issue-17602@issue-17602@systemd-update-utmp.service.service.service that would lead to infinite recursion
issue-17602@auditd.service.service: Dropping dependency issue-17602@issue-17602@auditd.service.service.service that would lead to infinite recursion
issue-17602@systemd-tmpfiles-setup.service.service: Dropping dependency issue-17602@issue-17602@systemd-tmpfiles-setup.service.service.service that would lead to infinite recursion
issue-17602@systemd-logind.service.service: Dropping dependency issue-17602@issue-17602@systemd-logind.service.service.service that would lead to infinite recursion
issue-17602@modprobe@drm.service.service: Dropping dependency issue-17602@issue-17602@modprobe@drm.service.service.service that would lead to infinite recursion
issue-17602@systemd-ask-password-wall.service.service: Dropping dependency issue-17602@issue-17602@systemd-ask-password-wall.service.service.service that would lead to infinite recursion
issue-17602@sshd.service.service: Dropping dependency issue-17602@issue-17602@sshd.service.service.service that would lead to infinite recursion
issue-17602@sshdgenkeys.service.service: Dropping dependency issue-17602@issue-17602@sshdgenkeys.service.service.service that would lead to infinite recursion
issue-17602@getty@tty1.service.service: Dropping dependency issue-17602@issue-17602@getty@tty1.service.service.service that would lead to infinite recursion
issue-17602@rc-local.service.service: Dropping dependency issue-17602@issue-17602@rc-local.service.service.service that would lead to infinite recursion
issue-17602@console-getty.service.service: Dropping dependency issue-17602@issue-17602@console-getty.service.service.service that would lead to infinite recursion
issue-17602@dbus.service.service: Dropping dependency issue-17602@issue-17602@dbus.service.service.service that would lead to infinite recursion
issue-17602@systemd-sysctl.service.service: Dropping dependency issue-17602@issue-17602@systemd-sysctl.service.service.service that would lead to infinite recursion
issue-17602@systemd-modules-load.service.service: Dropping dependency issue-17602@issue-17602@systemd-modules-load.service.service.service that would lead to infinite recursion
issue-17602@systemd-initctl.service.service: Dropping dependency issue-17602@issue-17602@systemd-initctl.service.service.service that would lead to infinite recursion
issue-17602@dm-event.service.service: Dropping dependency issue-17602@issue-17602@dm-event.service.service.service that would lead to infinite recursion
issue-17602@systemd-udev-trigger.service.service: Dropping dependency issue-17602@issue-17602@systemd-udev-trigger.service.service.service that would lead to infinite recursion
issue-17602@systemd-tmpfiles-setup-dev.service.service: Dropping dependency issue-17602@issue-17602@systemd-tmpfiles-setup-dev.service.service.service that would lead to infinite recursion
issue-17602@systemd-repart.service.service: Dropping dependency issue-17602@issue-17602@systemd-repart.service.service.service that would lead to infinite recursion
issue-17602@systemd-random-seed.service.service: Dropping dependency issue-17602@issue-17602@systemd-random-seed.service.service.service that would lead to infinite recursion
issue-17602@systemd-machine-id-commit.service.service: Dropping dependency issue-17602@issue-17602@systemd-machine-id-commit.service.service.service that would lead to infinite recursion
issue-17602@systemd-journal-flush.service.service: Dropping dependency issue-17602@issue-17602@systemd-journal-flush.service.service.service that would lead to infinite recursion
issue-17602@systemd-journal-catalog-update.service.service: Dropping dependency issue-17602@issue-17602@systemd-journal-catalog-update.service.service.service that would lead to infinite recursion
issue-17602@systemd-firstboot.service.service: Dropping dependency issue-17602@issue-17602@systemd-firstboot.service.service.service that would lead to infinite recursion
issue-17602@systemd-vconsole-setup.service.service: Dropping dependency issue-17602@issue-17602@systemd-vconsole-setup.service.service.service that would lead to infinite recursion
issue-17602@initrd-cleanup.service.service: Dropping dependency issue-17602@issue-17602@initrd-cleanup.service.service.service that would lead to infinite recursion
issue-17602@dracut-pre-udev.service.service: Dropping dependency issue-17602@issue-17602@dracut-pre-udev.service.service.service that would lead to infinite recursion
issue-17602@dracut-pre-trigger.service.service: Dropping dependency issue-17602@issue-17602@dracut-pre-trigger.service.service.service that would lead to infinite recursion
issue-17602@dracut-pre-pivot.service.service: Dropping dependency issue-17602@issue-17602@dracut-pre-pivot.service.service.service that would lead to infinite recursion
issue-17602@dracut-pre-mount.service.service: Dropping dependency issue-17602@issue-17602@dracut-pre-mount.service.service.service that would lead to infinite recursion
issue-17602@dracut-mount.service.service: Dropping dependency issue-17602@issue-17602@dracut-mount.service.service.service that would lead to infinite recursion
issue-17602@dracut-initqueue.service.service: Dropping dependency issue-17602@issue-17602@dracut-initqueue.service.service.service that would lead to infinite recursion
issue-17602@dracut-cmdline.service.service: Dropping dependency issue-17602@issue-17602@dracut-cmdline.service.service.service that would lead to infinite recursion
issue-17602@initrd-parse-etc.service.service: Dropping dependency issue-17602@issue-17602@initrd-parse-etc.service.service.service that would lead to infinite recursion
issue-17602@initrd-udevadm-cleanup-db.service.service: Dropping dependency issue-17602@issue-17602@initrd-udevadm-cleanup-db.service.service.service that would lead to infinite recursion
issue-17602@systemd-udev-settle.service.service: Dropping dependency issue-17602@issue-17602@systemd-udev-settle.service.service.service that would lead to infinite recursion
issue-17602@systemd-boot-system-token.service.service: Dropping dependency issue-17602@issue-17602@systemd-boot-system-token.service.service.service that would lead to infinite recursion
issue-17602@systemd-binfmt.service.service: Dropping dependency issue-17602@issue-17602@systemd-binfmt.service.service.service that would lead to infinite recursion
issue-17602@systemd-ask-password-console.service.service: Dropping dependency issue-17602@issue-17602@systemd-ask-password-console.service.service.service that would lead to infinite recursion
issue-17602@modprobe@configfs.service.service: Dropping dependency issue-17602@issue-17602@modprobe@configfs.service.service.service that would lead to infinite recursion
issue-17602@modprobe@fuse.service.service: Dropping dependency issue-17602@issue-17602@modprobe@fuse.service.service.service that would lead to infinite recursion
issue-17602@ldconfig.service.service: Dropping dependency issue-17602@issue-17602@ldconfig.service.service.service that would lead to infinite recursion
issue-17602@kmod-static-nodes.service.service: Dropping dependency issue-17602@issue-17602@kmod-static-nodes.service.service.service that would lead to infinite recursion
issue-17602@dracut-shutdown.service.service: Dropping dependency issue-17602@issue-17602@dracut-shutdown.service.service.service that would lead to infinite recursion
Queued start job for default target Graphical Interface.
[  OK  ] Created slice Slice /system/getty.
[  OK  ] Created slice Slice /system/modprobe.
[  OK  ] Created slice User and Session Slice.
[  OK  ] Started Dispatch Password Requests to Console Directory Watch.
[  OK  ] Started Forward Password Requests to Wall Directory Watch.
[  OK  ] Reached target Local Encrypted Volumes.
```

### CODE REVIEW

File: src/core/unit.c [hunk 1]
```diff
@@ -3180,6 +3180,52 @@ static int resolve_template(Unit *u, const char *name, char **buf, const char **
...
+        if (r < 0)
+                return r;
+
+        if (!u->instance || !instance)
+                return 0;
+
+        if (!strstr(instance, u->instance))
+                return 0;
+
+        log_unit_warning(u, "Dropping dependency %s that would lead to infinite recursion", name);
```
Comment (bluca) [inline src/core/unit.c]: maybe this should move to the caller? since this function is not really doing the dropping, it's just checking

File: src/core/unit.c [hunk 2]
```diff
@@ -3180,6 +3180,52 @@ static int resolve_template(Unit *u, const char *name, char **buf, const char **
...
+        int r;
+
+        /* If a template unit has a direct dependency on itself that contains the unit name as part of the
+         * template instance, this will almost certainly lead to infinite recursion as systemd will keep
+         * instantiating new instances of the template unit. https://github.com/systemd/systemd/issues/17602
+         * shows a good example of how this can happen in practice. To guard against this, we check for
+         * templates that depend on themselves and have the instantiated unit name as part of the template
+         * instance of the dependency.
+         */
+
```
Comment (bluca) [inline src/core/unit.c]: `assert(u)` (and maybe on `name` too?)

File: src/basic/unit-name.c [hunk 3]
```diff
@@ -797,3 +797,17 @@ bool slice_name_is_valid(const char *name) {
 
         return true;
 }
+
+bool unit_name_prefix_equal(const char *a, const char *b) {
+        const char *p, *q;
+
+        p = strchr(a, '@');
```
Comment (bluca) [inline src/basic/unit-name.c:811]: strchr bombs out if the string is null - hence either assert or check and return false if one/both are null

Comment (daandemeyer) [reply src/basic/unit-name.c:811]: Added asserts

File: src/basic/unit-name.c [hunk 4]
```diff
@@ -797,3 +797,20 @@ bool slice_name_is_valid(const char *name) {
...
+        if (!p)
+                p = strrchr(a, '.');
+
+        q = strchr(b, '@');
+        if (!q)
+                q = strrchr(b, '.');
+
+        return strneq(a, b, p - a);
```
Comment (keszybz) [inline src/basic/unit-name.c]: Hmm, `q` is not used for anything, and also `p` may be NULL if we're passed an invalid unit name.

File: src/core/unit.c [hunk 5]
```diff
@@ -3195,6 +3243,14 @@ int unit_add_dependency_by_name(Unit *u, UnitDependency d, const char *name, boo
         if (u->manager && FLAGS_SET(u->manager->test_run_flags, MANAGER_TEST_RUN_IGNORE_DEPENDENCIES))
                 return 0;
 
+        r = unit_is_recursive_template_dependency(u, name);
+        if (r < 0)
+                return r;
+        if (r == 1) {
+                log_unit_warning(u, "Dropping dependency %s that would lead to infinite recursion", name);
```
Comment (keszybz) [inline src/core/unit.c]: Maybe change "would lead to" to "could lead to" or even "raises suspicion of infinite recursion"? By saying "would" we open ourselves to critique that we got the condition wrong, but also we could confuse the user if there is in fact no recursion.

**Review by keszybz — CHANGES_REQUESTED** [review]
I think this is mostly OK (modulo some comments below). The test is very "heuristicky", but I don't think this will matter in practice. And it fixes a real issue. If people find some cases where we are too eager with rejecting dependencies, we can just adjust the test without too much work.

File: src/basic/unit-name.c (hunk 4 above)
Comment (daandemeyer) [reply src/basic/unit-name.c]: I added asserts so we require a valid unit name to be passed. Let le know if this is the way to go.

File: src/basic/unit-name.c [hunk 6]
```diff
@@ -797,3 +797,23 @@ bool slice_name_is_valid(const char *name) {
...
+                p = strrchr(a, '.');
+
+        q = strchr(b, '@');
+        if (!q)
+                q = strrchr(b, '.');
+
+        assert(p);
+        assert(q);
+
+        return p - a == q - b && strneq(a, b, p - a);
```
Comment (poettering) [inline src/basic/unit-name.c]: maybe:

```c
return memcmp_nn(a, p - a, b, q - b) == 0;
```

File: src/core/unit.c [hunk 7]
```diff
@@ -3195,6 +3243,14 @@ int unit_add_dependency_by_name(Unit *u, UnitDependency d, const char *name, boo
         if (u->manager && FLAGS_SET(u->manager->test_run_flags, MANAGER_TEST_RUN_IGNORE_DEPENDENCIES))
                 return 0;
 
+        r = unit_is_recursive_template_dependency(u, name);
+        if (r < 0)
+                return r;
+        if (r == 1) {
```
Comment (poettering) [inline src/core/unit.c]: i think it would be slightly nicer to check for `> 0` rather than `== 1` here, in the usual sense we have in place for tri-state bools (which this is not, but conceptually close)

File: src/core/unit.c [hunk 8]
```diff
@@ -3195,6 +3243,14 @@ int unit_add_dependency_by_name(Unit *u, UnitDependency d, const char *name, boo
...
+                log_unit_warning(u, "Dropping dependency %s that could lead to infinite recursion", name);
```
Comment (poettering) [inline src/core/unit.c]: maybe say "that likely leads to"...

Comment (poettering) [reply src/core/unit.c]: also end the sentence in full stop

Comment (poettering) [reply src/core/unit.c]: maybe also mention the dependency type, to make grepping for it easier, i.e.

```c
log_unit_warning(u, "Suppressing dependency %s=%s that likely leads to infinite recursion of instance templating.", unit_dependency_to_string(d), name);
```

File: src/core/unit.c [hunk 9]
```diff
@@ -3180,6 +3180,54 @@ static int resolve_template(Unit *u, const char *name, char **buf, const char **
...
+                return 0;
+
+        if (u->type != unit_name_to_type(name))
+                return 0;
+
+        r = unit_file_find_fragment(u->manager->unit_id_map, u->manager->unit_name_map, name, &fragment_path, &names);
+        if (r < 0)
+                return r;
+
+        if (!u->fragment_path || !fragment_path)
```
Comment (poettering) [inline src/core/unit.c]: smells as if the first half of this check should be moved before the unit_file_find_fragment() invocaiton, no?

File: src/core/unit.c [hunk 10]
```diff
@@ -3180,6 +3180,54 @@ static int resolve_template(Unit *u, const char *name, char **buf, const char **
...
+                return 0;
+
+        /* Fragment paths should also be equal as a custom fragment for a specific template instance
+         * wouldn't necessarily lead to infinite recursion. */
+        if (!streq(u->fragment_path, fragment_path))
```
Comment (poettering) [inline src/core/unit.c]: path_equal()

File: src/core/unit.c [hunk 11]
```diff
@@ -3180,6 +3180,54 @@ static int resolve_template(Unit *u, const char *name, char **buf, const char **
...
+        assert(name);
+
+        /* If a template unit has a direct dependency on itself that contains the unit name as part of the
+         * template instance, this will almost certainly lead to infinite recursion as systemd will keep
+         * instantiating new instances of the template unit. https://github.com/systemd/systemd/issues/17602
+         * shows a good example of how this can happen in practice. To guard against this, we check for
+         * templates that depend on themselves and have the instantiated unit name as part of the template
+         * instance of the dependency.
+         */
+
```
Comment (poettering) [inline src/core/unit.c]: hmm, there should be an early check here if this is a template at all, no?

File: src/core/unit.c [hunk 12]
```diff
@@ -3180,6 +3180,54 @@ static int resolve_template(Unit *u, const char *name, char **buf, const char **
...
+                return 0;
+
+        r = unit_name_to_instance(name, &instance);
+        if (r < 0)
+                return r;
+
+        if (!u->instance || !instance)
+                return 0;
+
+        if (!strstr(instance, u->instance))
```
Comment (poettering) [inline src/core/unit.c]: hmm, this is black magic. why not streq()?

**Review by poettering — CHANGES_REQUESTED** [review]

File: src/core/unit.c (hunk 12 above)
Comment (DaanDeMeyer) [reply src/core/unit.c:3225]: In the example in the linked issue, the global drop-in looks as follows:

```
[Unit]
Wants=notify@%n.service
```

Here the instance will be the full unit name, so `streq` would fail, but this still leads to infinite recursion as the full unit name contains the instance. That's why we use `strstr`, to catch any string that contains the instance, since all of those will almost surely lead to infinite recursion (If only the instance is sufficient to get recursion, every string containing the instance will surely do so as well).

**Comment by DaanDeMeyer** [comment]
Reworked the PR to check for specifiers in the dependency instance that expand to a superset of the instance (%i with extra characters, %n, %N). This seems like a better approach as specifiers are a requirement for infinite recursion to occur which we didn't check before.

File: src/core/load-fragment.c [hunk 13]
```diff
@@ -188,6 +263,16 @@ int config_parse_unit_deps(
                         continue;
                 }
 
+                r = unit_is_recursive_template_dependency(u, k, word);
+                if (r < 0)
+                        return r;
+                if (r > 0) {
+                        log_syntax(unit, LOG_WARNING, filename, line, r,
```
Comment (bluca) [inline src/core/load-fragment.c]: if this is a valid use case, then this shouldn't be a warning - better to document the dropping behaviour, and make this a debug or info at most

**Comment by bluca** [comment]
would it be possible to have a test for this?

**Comment by bluca** [comment]
you need a rebase to fix the openssl packit failure

File: src/core/load-fragment.c (hunk 13 above)
Comment (keszybz) [reply src/core/load-fragment.c]: No, it should be a warning. People should never depend on our hack to prevent recursion.

Comment (bluca) [reply src/core/load-fragment.c]: It is a valid use case though, for example the system level onfailure drop-in

Comment (keszybz) [reply src/core/load-fragment.c]: So… if it is a valid case, there should be no warning and no message and no drop. I think this will become clearer once we have the tests.

Comment (bluca) [reply src/core/load-fragment.c]: The valid use case needs the drop to work, see https://github.com/systemd/systemd/issues/17602

Comment (bluca) [reply src/core/load-fragment.c]: IE, the use case is adding this global drop-in to monitor failures:

```
/usr/lib/systemd/system/service.d/20-all-services-on-failure.conf:
    
[Unit]
OnFailure=failure-handler@%N.service
```

Which then causes a recursion. In other words, without workarounds like adding a symlink to /dev/null under failure-handler@.service.d/ to manually break it, it's not possible to reference instanced units in a top-level global drop-in.

Comment (keszybz) [reply src/core/load-fragment.c]: OK, I see what you mean. But then the docs should be adjusted to list the cases where we officially stop recursion. The heuristic can catch more cases. But the clear ones should be documented and then they should be dropped silently.

Comment (bluca) [reply src/core/load-fragment.c]: yes, documentation needs an update

Comment (daandemeyer) [reply src/core/load-fragment.c]: How would we detect a valid use case though? Maybe we just drop this to debug and document that we stop recursion?

Comment (daandemeyer) [reply src/core/load-fragment.c]: Dropped to debug and modified the docs to mention we automatically remove direct and recursive dependencies from template units on themselves.

File: man/systemd.unit.xml [hunk 14]
```diff
@@ -2256,9 +2256,9 @@ OnFailure=failure-handler@%N.service
     services will have acquired an <varname>OnFailure=</varname> dependency on
     <filename index='false'>failure-handler@%N.service</filename>. The
     template instance units will also have gained the dependency which results
-    in the creation of a recursive dependency chain. We can break the chain by
-    disabling the drop-in for the template instance units via a symlink to
-    <filename index='false'>/dev/null</filename>:</para>
```
Comment (bluca) [inline man/systemd.unit.xml:2261]: This is a useful tip, so I'd leave it there, maybe reworded

File: src/test/test-load-fragment.c [hunk 15]
```diff
@@ -850,6 +891,6 @@ int main(int argc, char *argv[]) {
         TEST_REQ_RUNNING_SYSTEMD(test_install_printf());
         test_unit_dump_config_items();
         test_config_parse_memory_limit();
-
+        test_unit_is_recursive_template_dependency();
```
Comment (keszybz) [inline src/test/test-load-fragment.c:919]: Please keep the empty line.

File: src/test/test-load-fragment.c [hunk 16]
```diff
@@ -829,6 +829,47 @@ static void test_config_parse_memory_limit(void) {
...
+                return;
+        }
+
+        assert_se(r >= 0);
+        assert_se(manager_startup(m, NULL, NULL, NULL) >= 0);
+
+        assert_se(u = unit_new(m, sizeof(Service)));
+        u->type = UNIT_SERVICE;
+        u->id = (char[]) { "foobar@.service" };
+        u->fragment_path = (char[]) { "/foobar@.service" };
```
Comment (keszybz) [inline src/test/test-load-fragment.c]: This is rather strange… `.id` and `.fragment_path` are `char*`, and this assigns a `char*[]` array to them. Also, the unit is not freed automatically by the manager, but manually, which is fragile. Please see how e.g. `test-unit-name` does this.

File: src/core/load-fragment.c [hunk 17]
```diff
@@ -150,6 +150,85 @@ DEFINE_CONFIG_PARSE_ENUM_WITH_DEFAULT(config_parse_numa_policy, mpol, int, -1, "
...
+                if (*p == '%')
+                        percent = !percent;
+                else if (percent && (*p == 'n' || *p == 'N' || *p == 'i'))
+                        return true;
+        }
+
+        return false;
+}
+
+int unit_is_recursive_template_dependency(Unit *u, const char *name, const char *format) {
```
Comment (keszybz) [inline src/core/load-fragment.c]: Maybe add a comment here that `name` is `format` rendered.

File: src/core/load-fragment.c [hunk 18]
```diff
@@ -150,6 +150,85 @@ DEFINE_CONFIG_PARSE_ENUM_WITH_DEFAULT(config_parse_numa_policy, mpol, int, -1, "
...
+
+        if (!u->fragment_path)
+                return 0;
+
+        r = unit_file_find_fragment(u->manager->unit_id_map, u->manager->unit_name_map, name, &fragment_path, &names);
+        if (r < 0)
+                return r;
+
+        if (!fragment_path)
+                return 0;
```
Comment (keszybz) [inline src/core/load-fragment.c]: Not everything has a fragment, for example .slice and .device units. Shouldn't this check be removed, and the check below amended to treat NULL as just another distinct value of `fragment_path` ?

File: src/core/load-fragment.c [hunk 19]
```diff
@@ -150,6 +150,85 @@ DEFINE_CONFIG_PARSE_ENUM_WITH_DEFAULT(config_parse_numa_policy, mpol, int, -1, "
...
+
+        /* If the string is just the instance specifier, it's not a superset of the instance. */
+        if (memcmp_nn(p, q - p, "%i", strlen("%i")) == 0)
+                return false;
+
+        /* %i, %n and %N all expand to the instance or a superset of it. */
+        for (; *p != '\0' && *p != '.'; p++) {
+                if (*p == '%')
+                        percent = !percent;
+                else if (percent && (*p == 'n' || *p == 'N' || *p == 'i'))
```
Comment (keszybz) [inline src/core/load-fragment.c]: `IN_SET(*p, 'n', 'N', 'i')`

I think this cannot work because `percent` is not set to false again.
IIUC, this will return false for "%x%n", but true for "%xn". 
Please add those test cases to the test.

File: man/systemd.unit.xml (hunk 14 above)
Comment (daandemeyer) [reply man/systemd.unit.xml:2261]: Is it still necessary though since systemd automatically breaks the dependency now? (admittedly not in all cases)

Comment (bluca) [reply man/systemd.unit.xml:2261]: In many cases no, so it can be reworded as a fallback, but it's a non-obvious solution so it's worth keeping it documented

File: src/core/load-fragment.c [hunk 20]
```diff
@@ -150,6 +150,83 @@ DEFINE_CONFIG_PARSE_ENUM_WITH_DEFAULT(config_parse_numa_policy, mpol, int, -1, "
...
+        const char *p, *q;
+        bool percent = false;
+
+        assert(s);
+
+        p = strchr(s, '@');
+        if (!p)
+                return false;
+
+        p++; // Skip '@'
```
Comment (poettering) [inline src/core/load-fragment.c]: in our codebase we generally want to use `//` only for "fixme" style comments and `/**/` for "explanatory" comments. This is the latter, hence please use /* */

(Also see coding style docs)

File: src/core/load-fragment.c [hunk 21]
```diff
@@ -150,6 +150,83 @@ DEFINE_CONFIG_PARSE_ENUM_WITH_DEFAULT(config_parse_numa_policy, mpol, int, -1, "
...
+        if (!unit_name_is_valid(name, UNIT_NAME_INSTANCE))
+                return 0;
+
+        if (!unit_name_prefix_equal(u->id, name))
+                return 0;
+
+        if (u->type != unit_name_to_type(name))
+                return 0;
+
+        r = unit_file_find_fragment(u->manager->unit_id_map, u->manager->unit_name_map, name, &fragment_path, &names);
```
Comment (poettering) [inline src/core/load-fragment.c]: you are not actually interested in the `names` return parameter. Thus I think it would be better to fix unit_file_find_fragment() to make that param optional, i.e. that if we pass NULL there it just doesn't return it and doesn't bother with collecting that info

File: src/core/load-fragment.c [hunk 22]
```diff
@@ -150,6 +150,83 @@ DEFINE_CONFIG_PARSE_ENUM_WITH_DEFAULT(config_parse_numa_policy, mpol, int, -1, "
...
+
+        assert(s);
+
+        p = strchr(s, '@');
+        if (!p)
+                return false;
+
+        p++; // Skip '@'
+
+        q = strchrnul(p, '.');
```
Comment (poettering) [inline src/core/load-fragment.c]: this should probably look for the last `.`, not the first

File: src/core/load-fragment.c [hunk 23]
```diff
@@ -150,6 +150,83 @@ DEFINE_CONFIG_PARSE_ENUM_WITH_DEFAULT(config_parse_numa_policy, mpol, int, -1, "
...
+
+        /* If the string is just the instance specifier, it's not a superset of the instance. */
+        if (memcmp_nn(p, q - p, "%i", strlen("%i")) == 0)
+                return false;
+
+        /* %i, %n and %N all expand to the instance or a superset of it. */
+        for (; *p != '\0' && *p != '.'; p++) {
```
Comment (poettering) [inline src/core/load-fragment.c]: condition check `p < q` here, no?

File: src/core/load-fragment.c [hunk 24]
```diff
@@ -150,6 +150,83 @@ DEFINE_CONFIG_PARSE_ENUM_WITH_DEFAULT(config_parse_numa_policy, mpol, int, -1, "
...
+         * practice. To guard against this, we check for templates that depend on themselves and have the
+         * instantiated unit instance included as part of the template instance of the dependency via a
+         * specifier.
+         *
+         * For example, if systemd-notify@.service depends on systemd-notify@%n.service, this will result in
+         * infinite recursion.
+         */
+
+        if (!unit_name_is_valid(name, UNIT_NAME_INSTANCE))
+                return 0;
```
Comment (poettering) [inline src/core/load-fragment.c]: given this function returns a boolean-like return code (i.e. basically true, false, or error), i'd write `return false` here, to make that even clearer. (and below too)

File: src/core/load-fragment.c [hunk 25]
```diff
@@ -189,6 +266,16 @@ int config_parse_unit_deps(
                         continue;
                 }
 
+                r = unit_is_recursive_template_dependency(u, k, word);
+                if (r < 0)
+                        return r;
+                if (r > 0) {
+                        log_syntax(unit, LOG_DEBUG, filename, line, r,
```
Comment (poettering) [inline src/core/load-fragment.c]: `r` does not contain any error code here

File: src/core/load-fragment.c [hunk 26]
```diff
@@ -189,6 +266,16 @@ int config_parse_unit_deps(
                         continue;
                 }
 
+                r = unit_is_recursive_template_dependency(u, k, word);
+                if (r < 0)
+                        return r;
```
Comment (poettering) [inline src/core/load-fragment.c]: should probably log the error

File: src/core/load-fragment.c [hunk 27]
```diff
@@ -150,6 +150,83 @@ DEFINE_CONFIG_PARSE_ENUM_WITH_DEFAULT(config_parse_numa_policy, mpol, int, -1, "
...
+                                return true;
+                        percent = false;
+                }
+        }
+
+        return false;
+}
+
+/* `name` is the rendered version of `format` via `unit_printf` or similar functions. */
+int unit_is_recursive_template_dependency(Unit *u, const char *name, const char *format) {
```
Comment (poettering) [inline src/core/load-fragment.c]: this is not giving 100% accurate asnwers, right? maybe rename `unit_is_likely_recursive_dependency()`?

File: src/basic/unit-name.c [hunk 28]
```diff
@@ -797,3 +798,23 @@ bool slice_name_is_valid(const char *name) {
...
+
+        q = strchr(b, '@');
+        if (!q)
+                q = strrchr(b, '.');
+
+        assert(p);
+        assert(q);
+
+        return memcmp_nn(a, p - a, b, q - b) == 0;
+}
```
Comment (poettering) [inline src/basic/unit-name.c:823]: i'd prefer it his function would have a unit test of its own

File: src/core/load-fragment.c [hunk 29]
```diff
@@ -150,6 +150,83 @@ DEFINE_CONFIG_PARSE_ENUM_WITH_DEFAULT(config_parse_numa_policy, mpol, int, -1, "
...
+                        percent = !percent;
+                else if (percent) {
+                        if (IN_SET(*p, 'n', 'N', 'i'))
+                                return true;
+                        percent = false;
+                }
+        }
+
+        return false;
+}
```
Comment (poettering) [inline src/core/load-fragment.c:185]: unit test for this would be good too, it's not entirely clear to me if this does everything the way it should, and a unit test would certainly make any issues visible.
